In [2]:
import pandas as pd
pd . read_csv ( 'MailCustomer.csv' , index_col = 'CustomerKey' , sep = ';' , encoding = " ISO-8859 -1 " )

GeographyKey CustomerAlternateKey Title  FirstName MiddleName  \
CustomerKey                                                                  
11000                  26           AW00011000   NaN        Jon          V   
11001                  37           AW00011001   NaN     Eugene          L   
11002                  31           AW00011002   NaN      Ruben        NaN   
11003                  11           AW00011003   NaN    Christy        NaN   
11004                  19           AW00011004   NaN  Elizabeth        NaN   
...                   ...                  ...   ...        ...        ...   
29479                 209           AW00029479   NaN      Tommy          L   
29480                 248           AW00029480   NaN       Nina          W   
29481                 120           AW00029481   NaN       Ivan        NaN   
29482                 179           AW00029482   NaN    Clayton        NaN   
29483                 217           AW00029483   NaN      J‚sus          L   

            LastName  NameStyle  BirthDate MaritalStatus Suffix  ...  \
CustomerKey                                                      ...   
11000           Yang          0     4/8/66             M    NaN  ...   
11001          Huang          0  5/14/1965             S    NaN  ...   
11002         Torres          0    8/12/65             M    NaN  ...   
11003            Zhu          0  2/15/1968             S    NaN  ...   
11004        Johnson          0     8/8/68             S    NaN  ...   
...              ...        ...        ...           ...    ...  ...   
29479           Tang          0     7/4/58             M    NaN  ...   
29480           Raji          0   11/10/60             S    NaN  ...   
29481           Suri          0     1/5/60             S    NaN  ...   
29482          Zhang          0     3/5/59             M    NaN  ...   
29483        Navarro          0    12/8/59             M    NaN  ...   

            HouseOwnerFlag NumberCarsOwned               AddressLine1  \
CustomerKey                                                             
11000                    1               0            3761 N. 14th St   
11001                    0               1                 2243 W St.   
11002                    1               1           5844 Linden Land   
11003                    0               1           1825 Village Pl.   
11004                    1               4        7553 Harness Circle   
...                    ...             ...                        ...   
29479                    1               0          111, rue Maillard   
29480                    1               0          9 Katherine Drive   
29481                    0               0                Knaackstr 4   
29482                    1               0     1080, quai de Grenelle   
29483                    1               0  244, rue de la Centenaire   

             AddressLine2                Phone DateFirstPurchase  \
CustomerKey                                                        
11000                 NaN  1 (11) 500 555-0162         7/22/2005   
11001                 NaN  1 (11) 500 555-0110         7/18/2005   
11002                 NaN  1 (11) 500 555-0184           7/10/05   
11003                 NaN  1 (11) 500 555-0162            7/1/05   
11004                 NaN  1 (11) 500 555-0131         7/26/2005   
...                   ...                  ...               ...   
29479                 NaN  1 (11) 500 555-0136            3/8/07   
29480                 NaN  1 (11) 500 555-0146         1/18/2008   
29481                 NaN  1 (11) 500 555-0144         2/13/2006   
29482                 NaN  1 (11) 500 555-0137         3/22/2007   
29483                 NaN  1 (11) 500 555-0141         3/13/2007   

            CommuteDistance   Region Age BikeBuyer  
CustomerKey                                         
11000             1-2 Miles  Pacific  49         1  
11001             0-1 Miles  Pacific  50         1  
11002         

In [3]:
data = pd.read_csv ( 'MailCustomer.csv' , index_col = 'CustomerKey' , sep = ';' , encoding= " ISO -8859 -1 ")
print (data.head() )
print ( " - - - - - - - - - - - - - - - - - - - - ")
print (data.tail() )

data.tail()

             GeographyKey CustomerAlternateKey Title  FirstName MiddleName  \
CustomerKey                                                                  
11000                  26           AW00011000   NaN        Jon          V   
11001                  37           AW00011001   NaN     Eugene          L   
11002                  31           AW00011002   NaN      Ruben        NaN   
11003                  11           AW00011003   NaN    Christy        NaN   
11004                  19           AW00011004   NaN  Elizabeth        NaN   

            LastName  NameStyle  BirthDate MaritalStatus Suffix  ...  \
CustomerKey                                                      ...   
11000           Yang          0     4/8/66             M    NaN  ...   
11001          Huang          0  5/14/1965             S    NaN  ...   
11002         Torres          0    8/12/65             M    NaN  ...   
11003            Zhu          0  2/15/1968             S    NaN  ...   
11004        Johnson 

GeographyKey CustomerAlternateKey Title FirstName MiddleName  \
CustomerKey                                                                 
29479                 209           AW00029479   NaN     Tommy          L   
29480                 248           AW00029480   NaN      Nina          W   
29481                 120           AW00029481   NaN      Ivan        NaN   
29482                 179           AW00029482   NaN   Clayton        NaN   
29483                 217           AW00029483   NaN     J‚sus          L   

            LastName  NameStyle BirthDate MaritalStatus Suffix  ...  \
CustomerKey                                                     ...   
29479           Tang          0    7/4/58             M    NaN  ...   
29480           Raji          0  11/10/60             S    NaN  ...   
29481           Suri          0    1/5/60             S    NaN  ...   
29482          Zhang          0    3/5/59             M    NaN  ...   
29483        Navarro          0   12/8/59             M    NaN  ...   

            HouseOwnerFlag NumberCarsOwned               AddressLine1  \
CustomerKey                                                             
29479                    1               0          111, rue Maillard   
29480                    1               0          9 Katherine Drive   
29481                    0               0                Knaackstr 4   
29482                    1               0     1080, quai de Grenelle   
29483                    1               0  244, rue de la Centenaire   

             AddressLine2                Phone DateFirstPurchase  \
CustomerKey                                                        
29479                 NaN  1 (11) 500 555-0136            3/8/07   
29480                 NaN  1 (11) 500 555-0146         1/18/2008   
29481                 NaN  1 (11) 500 555-0144         2/13/2006   
29482                 NaN  1 (11) 500 555-0137         3/22/2007   
29483                 NaN  1 (11) 500 555-0141         3/13/2007   

            CommuteDistance  Region Age BikeBuyer  
CustomerKey                                        
29479             0-1 Miles  Europe  57         1  
29480             0-1 Miles  Europe  54         1  
29481             0-1 Miles  Europe  55         1  
29482             0-1 Miles  Europe  56         1  
29483             0-1 Miles  Europe  55         1  

[5 rows x 31 columns]

In [4]:
print(data.iloc[0]) # Muestra la primera fila
print(data.iloc[: ,0]) # Muestra la primera columna
print(data.iloc[0:5]) # Primeras cinco filas
print(data.iloc[: , 0:5]) # Primeras cinco columnas
print(data.iloc[[0 ,2 ,1]]) # Primera , tercera y segunda filas

GeographyKey                                   26
CustomerAlternateKey                   AW00011000
Title                                         NaN
FirstName                                     Jon
MiddleName                                      V
LastName                                     Yang
NameStyle                                       0
BirthDate                                  4/8/66
MaritalStatus                                   M
Suffix                                        NaN
Gender                                          M
EmailAddress            jon24@adventure-works.com
YearlyIncome                                90000
TotalChildren                                   2
NumberChildrenAtHome                            0
EnglishEducation                        Bachelors
SpanishEducation                     Licenciatura
FrenchEducation                           Bac + 4
EnglishOccupation                    Professional
SpanishOccupation                     Profesional


In [5]:
print(data.columns) # Lista con los nombres de las columnas
print(data.info()) # Informacion util de los datos , como el tipo de datos

Index(['GeographyKey', 'CustomerAlternateKey', 'Title', 'FirstName',
       'MiddleName', 'LastName', 'NameStyle', 'BirthDate', 'MaritalStatus',
       'Suffix', 'Gender', 'EmailAddress', 'YearlyIncome', 'TotalChildren',
       'NumberChildrenAtHome', 'EnglishEducation', 'SpanishEducation',
       'FrenchEducation', 'EnglishOccupation', 'SpanishOccupation',
       'FrenchOccupation', 'HouseOwnerFlag', 'NumberCarsOwned', 'AddressLine1',
       'AddressLine2', 'Phone', 'DateFirstPurchase', 'CommuteDistance',
       'Region', 'Age', 'BikeBuyer'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
Index: 18484 entries, 11000 to 29483
Data columns (total 31 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   GeographyKey          18484 non-null  int64 
 1   CustomerAlternateKey  18484 non-null  object
 2   Title                 101 non-null    object
 3   FirstName             18484 non-null  object
 4   MiddleName     

In [6]:
print(data[[ 'FirstName' , 'LastName' ]]) # Acceso al nombre y apellido
print(data.loc[: , [ 'FirstName' , 'LastName' ]]) # Acceso al nombre y apellido
# Si queremos saber cuantos valores distintos tiene cada una de nuestras columnas, podemos hacer lo siguiente:
for i in data . columns :
  print (i , " -->" , len(data[i].unique().tolist()))

             FirstName LastName
CustomerKey                    
11000              Jon     Yang
11001           Eugene    Huang
11002            Ruben   Torres
11003          Christy      Zhu
11004        Elizabeth  Johnson
...                ...      ...
29479            Tommy     Tang
29480             Nina     Raji
29481             Ivan     Suri
29482          Clayton    Zhang
29483            J‚sus  Navarro

[18484 rows x 2 columns]
             FirstName LastName
CustomerKey                    
11000              Jon     Yang
11001           Eugene    Huang
11002            Ruben   Torres
11003          Christy      Zhu
11004        Elizabeth  Johnson
...                ...      ...
29479            Tommy     Tang
29480             Nina     Raji
29481             Ivan     Suri
29482          Clayton    Zhang
29483            J‚sus  Navarro

[18484 rows x 2 columns]
GeographyKey  --> 336
CustomerAlternateKey  --> 18484
Title  --> 6
FirstName  --> 670
MiddleName  --> 45
LastName  -

In [7]:
# Datos faltantes
print(data.isnull().sum().sort_values(ascending = False))

Suffix                  18481
Title                   18383
AddressLine2            18172
MiddleName               7830
GeographyKey                0
SpanishOccupation           0
FrenchOccupation            0
HouseOwnerFlag              0
NumberCarsOwned             0
AddressLine1                0
Phone                       0
FrenchEducation             0
DateFirstPurchase           0
CommuteDistance             0
Region                      0
Age                         0
EnglishOccupation           0
EnglishEducation            0
SpanishEducation            0
CustomerAlternateKey        0
NumberChildrenAtHome        0
TotalChildren               0
YearlyIncome                0
EmailAddress                0
Gender                      0
MaritalStatus               0
BirthDate                   0
NameStyle                   0
LastName                    0
FirstName                   0
BikeBuyer                   0
dtype: int64


In [8]:
processedData = data.fillna(" None " )
print(processedData.isnull().sum().sort_values(ascending = False ) ) # Comprobamos que ya no hay valores nulos

GeographyKey            0
SpanishEducation        0
Age                     0
Region                  0
CommuteDistance         0
DateFirstPurchase       0
Phone                   0
AddressLine2            0
AddressLine1            0
NumberCarsOwned         0
HouseOwnerFlag          0
FrenchOccupation        0
SpanishOccupation       0
EnglishOccupation       0
FrenchEducation         0
EnglishEducation        0
CustomerAlternateKey    0
NumberChildrenAtHome    0
TotalChildren           0
YearlyIncome            0
EmailAddress            0
Gender                  0
Suffix                  0
MaritalStatus           0
BirthDate               0
NameStyle               0
LastName                0
MiddleName              0
FirstName               0
Title                   0
BikeBuyer               0
dtype: int64


In [9]:
data.describe()

GeographyKey  NameStyle   YearlyIncome  TotalChildren  \
count  18484.000000    18484.0   18484.000000   18484.000000   
mean     257.956287        0.0   57305.777970       1.844352   
std      196.531062        0.0   32285.841703       1.612408   
min        2.000000        0.0   10000.000000       0.000000   
25%       62.000000        0.0   30000.000000       0.000000   
50%      240.000000        0.0   60000.000000       2.000000   
75%      345.000000        0.0   70000.000000       3.000000   
max      654.000000        0.0  170000.000000       5.000000   

       NumberChildrenAtHome  HouseOwnerFlag  NumberCarsOwned           Age  \
count          18484.000000    18484.000000     18484.000000  18484.000000   
mean               1.004058        0.676369         1.502705     53.135793   
std                1.522660        0.467874         1.138394     11.261340   
min                0.000000        0.000000         0.000000     34.000000   
25%                0.000000        0.000000         1.000000     44.000000   
50%                0.000000        1.000000         2.000000     52.000000   
75%                2.000000        1.000000         2.000000     61.000000   
max                5.000000        1.000000         4.000000    105.000000   

          BikeBuyer  
count  18484.000000  
mean       0.494049  
std        0.499978  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        1.000000  
max        1.000000

In [10]:
# Ver los tipos de datos
data.dtypes

GeographyKey             int64
CustomerAlternateKey    object
Title                   object
FirstName               object
MiddleName              object
LastName                object
NameStyle                int64
BirthDate               object
MaritalStatus           object
Suffix                  object
Gender                  object
EmailAddress            object
YearlyIncome             int64
TotalChildren            int64
NumberChildrenAtHome     int64
EnglishEducation        object
SpanishEducation        object
FrenchEducation         object
EnglishOccupation       object
SpanishOccupation       object
FrenchOccupation        object
HouseOwnerFlag           int64
NumberCarsOwned          int64
AddressLine1            object
AddressLine2            object
Phone                   object
DateFirstPurchase       object
CommuteDistance         object
Region                  object
Age                      int64
BikeBuyer                int64
dtype: object

In [11]:
data[data['BirthDate'].isnull()] # Ver donde la columna BirthDate es nula

Empty DataFrame
Columns: [GeographyKey, CustomerAlternateKey, Title, FirstName, MiddleName, LastName, NameStyle, BirthDate, MaritalStatus, Suffix, Gender, EmailAddress, YearlyIncome, TotalChildren, NumberChildrenAtHome, EnglishEducation, SpanishEducation, FrenchEducation, EnglishOccupation, SpanishOccupation, FrenchOccupation, HouseOwnerFlag, NumberCarsOwned, AddressLine1, AddressLine2, Phone, DateFirstPurchase, CommuteDistance, Region, Age, BikeBuyer]
Index: []

[0 rows x 31 columns]

In [12]:
# Ver datos únicos
for column in data.columns:
    print(f"Column '{column}' has {data[column].nunique()} unique values.")


Column 'GeographyKey' has 336 unique values.
Column 'CustomerAlternateKey' has 18484 unique values.
Column 'Title' has 5 unique values.
Column 'FirstName' has 670 unique values.
Column 'MiddleName' has 44 unique values.
Column 'LastName' has 375 unique values.
Column 'NameStyle' has 1 unique values.
Column 'BirthDate' has 8252 unique values.
Column 'MaritalStatus' has 2 unique values.
Column 'Suffix' has 1 unique values.
Column 'Gender' has 2 unique values.
Column 'EmailAddress' has 18484 unique values.
Column 'YearlyIncome' has 16 unique values.
Column 'TotalChildren' has 6 unique values.
Column 'NumberChildrenAtHome' has 6 unique values.
Column 'EnglishEducation' has 5 unique values.
Column 'SpanishEducation' has 5 unique values.
Column 'FrenchEducation' has 5 unique values.
Column 'EnglishOccupation' has 5 unique values.
Column 'SpanishOccupation' has 5 unique values.
Column 'FrenchOccupation' has 5 unique values.
Column 'HouseOwnerFlag' has 2 unique values.
Column 'NumberCarsOwned'

In [16]:
# Resumen:
# Obtener información básica del DataFrame

data_info = data.info()

data_head = data.head()

data_tail = data.tail()

data_nulls = data.isnull().sum().sort_values(ascending=False)



data_info, data_head, data_tail, data_nulls

<class 'pandas.core.frame.DataFrame'>
Index: 18484 entries, 11000 to 29483
Data columns (total 31 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   GeographyKey          18484 non-null  int64 
 1   CustomerAlternateKey  18484 non-null  object
 2   Title                 101 non-null    object
 3   FirstName             18484 non-null  object
 4   MiddleName            10654 non-null  object
 5   LastName              18484 non-null  object
 6   NameStyle             18484 non-null  int64 
 7   BirthDate             18484 non-null  object
 8   MaritalStatus         18484 non-null  object
 9   Suffix                3 non-null      object
 10  Gender                18484 non-null  object
 11  EmailAddress          18484 non-null  object
 12  YearlyIncome          18484 non-null  int64 
 13  TotalChildren         18484 non-null  int64 
 14  NumberChildrenAtHome  18484 non-null  int64 
 15  EnglishEducation      18484 non-null 

(None,
              GeographyKey CustomerAlternateKey Title  FirstName MiddleName  \
 CustomerKey                                                                  
 11000                  26           AW00011000   NaN        Jon          V   
 11001                  37           AW00011001   NaN     Eugene          L   
 11002                  31           AW00011002   NaN      Ruben        NaN   
 11003                  11           AW00011003   NaN    Christy        NaN   
 11004                  19           AW00011004   NaN  Elizabeth        NaN   
 
             LastName  NameStyle  BirthDate MaritalStatus Suffix  ...  \
 CustomerKey                                                      ...   
 11000           Yang          0     4/8/66             M    NaN  ...   
 11001          Huang          0  5/14/1965             S    NaN  ...   
 11002         Torres          0    8/12/65             M    NaN  ...   
 11003            Zhu          0  2/15/1968             S    NaN  ...   


1. Estructura del Conjunto de Datos

    Total de filas: 18,484
    Total de columnas: 31
    Tipos de datos:
        9 columnas son numéricas (int64).
        22 columnas son de tipo texto o categóricas (object).

2. Datos Faltantes

Los datos faltantes, las 3 columnas más problemáticas son:

    Suffix: 18,481 valores faltantes (99.98% de los datos).
    Title: 18,383 valores faltantes (99.45% de los datos).
    AddressLine2: 18,172 valores faltantes (98.31% de los datos).

Algunas columnas como MiddleName también tienen datos faltantes significativos (~42%).

3. Vista Preliminar de las Filas

Ejemplos de las primeras y últimas filas confirman que los datos incluyen información como nombres, direcciones, ingresos, estado civil, género, y edad. Algunas observaciones iniciales:

    La columna BirthDate tiene valores de fecha almacenados como cadenas, lo que podría requerir conversión.
    Columnas redundantes parecen estar presentes (e.g., EnglishEducation, SpanishEducation, FrenchEducation).